File: web/templates.js
/**
 * templates.js — Built-in diagram templates, mirroring templates.c
 */

const Templates = {
  loadDefaultFlowchart(diagram) {
    diagram.clear();

    const nStart = diagram.addNode(ShapeType.TERMINATOR, 80,  240, 130, 50, 'Iniciar Pedido');
    const nData  = diagram.addNode(ShapeType.DATA,       270, 235, 150, 60, 'Receber Dados');
    const nDec   = diagram.addNode(ShapeType.DECISION,   480, 215, 140, 100,'Estoque OK?');
    const nProc  = diagram.addNode(ShapeType.PROCESS,    680, 235, 140, 60, 'Reservar Itens');
    const nSub   = diagram.addNode(ShapeType.SUBPROCESS, 880, 235, 150, 60, 'Gateway Pagamento');
    const nDB    = diagram.addNode(ShapeType.DATABASE,   1090,225, 130, 80, 'Banco Postgres');
    const nEnd   = diagram.addNode(ShapeType.TERMINATOR, 1290,240, 140, 50, 'Pedido Entregue');

    const nWarn  = diagram.addNode(ShapeType.PROCESS,    480, 380, 150, 60, 'Avisar Cliente');
    if (nWarn) {
      nWarn.fillColor   = { r: 225, g: 29,  b: 72,  a: 230 };
      nWarn.borderColor = { r: 254, g: 205, b: 211, a: 255 };
    }

    const nFail  = diagram.addNode(ShapeType.TERMINATOR, 480, 490, 140, 50, 'Encerrar Pedido');
    if (nFail) {
      nFail.fillColor   = { r: 190, g: 18,  b: 60,  a: 230 };
      nFail.borderColor = { r: 253, g: 164, b: 175, a: 255 };
    }

    diagram.addNode(ShapeType.NOTE, 880, 360, 160, 80, 'Nota: Validação PIX e Cartão com webhook');

    // Connections
    if (nStart && nData)  diagram.addConnection(nStart.id, nData.id,  PortIndex.RIGHT, PortIndex.LEFT,   '');
    if (nData  && nDec)   diagram.addConnection(nData.id,  nDec.id,   PortIndex.RIGHT, PortIndex.LEFT,   '');
    if (nDec   && nProc)  diagram.addConnection(nDec.id,   nProc.id,  PortIndex.RIGHT, PortIndex.LEFT,   'Sim');
    if (nDec   && nWarn)  diagram.addConnection(nDec.id,   nWarn.id,  PortIndex.BOTTOM,PortIndex.TOP,    'Não');
    if (nWarn  && nFail)  diagram.addConnection(nWarn.id,  nFail.id,  PortIndex.BOTTOM,PortIndex.TOP,    '');
    if (nProc  && nSub)   diagram.addConnection(nProc.id,  nSub.id,   PortIndex.RIGHT, PortIndex.LEFT,   'Aprovado');
    if (nSub   && nDB)    diagram.addConnection(nSub.id,   nDB.id,    PortIndex.RIGHT, PortIndex.LEFT,   'Persistir');
    if (nDB    && nEnd)   diagram.addConnection(nDB.id,    nEnd.id,   PortIndex.RIGHT, PortIndex.LEFT,   'OK');

    diagram.selectedNodeId = -1;
    diagram.selectedConnId = -1;
    diagram.setToast('Exemplo de Fluxograma Carregado!');
  },

  loadSystemArchitecture(diagram) {
    diagram.clear();

    const nClient  = diagram.addNode(ShapeType.TERMINATOR, 100, 250, 140, 60,  'Frontend Web');
    const nGateway = diagram.addNode(ShapeType.PROCESS,    320, 250, 150, 60,  'API Gateway (C)');
    const nAuth    = diagram.addNode(ShapeType.PROCESS,    550, 150, 150, 60,  'Auth Service');
    const nCore    = diagram.addNode(ShapeType.SUBPROCESS, 550, 340, 160, 70,  'Core Engine');
    const nCache   = diagram.addNode(ShapeType.DATABASE,   780, 150, 130, 70,  'Redis Cache');
    const nDB      = diagram.addNode(ShapeType.DATABASE,   780, 340, 130, 80,  'Cluster SQL');

    if (nClient && nGateway) diagram.addConnection(nClient.id,  nGateway.id, PortIndex.RIGHT, PortIndex.LEFT,  'HTTPS/JSON');
    if (nGateway && nAuth)   diagram.addConnection(nGateway.id, nAuth.id,    PortIndex.RIGHT, PortIndex.LEFT,  'Validar JWT');
    if (nAuth && nCache)     diagram.addConnection(nAuth.id,    nCache.id,   PortIndex.RIGHT, PortIndex.LEFT,  'Sessão');
    if (nGateway && nCore)   diagram.addConnection(nGateway.id, nCore.id,    PortIndex.RIGHT, PortIndex.LEFT,  'gRPC Call');
    if (nCore && nDB)        diagram.addConnection(nCore.id,    nDB.id,      PortIndex.RIGHT, PortIndex.LEFT,  'Queries');

    diagram.selectedNodeId = -1;
    diagram.selectedConnId = -1;
    diagram.setToast('Arquitetura de Sistema Carregada!');
  },
};

File: src/templates.c
#include "templates.h"

void Templates_LoadDefaultFlowchart(Diagram *d) {
    Diagram_Clear(d);

    // 1. Início
    Node *nStart = Diagram_AddNode(d, SHAPE_TERMINATOR, 80, 240, 130, 50, "Iniciar Pedido");

    // 2. Entrada de Dados
    Node *nData = Diagram_AddNode(d, SHAPE_DATA, 270, 235, 150, 60, "Receber Dados");

    // 3. Decisão
    Node *nDec = Diagram_AddNode(d, SHAPE_DECISION, 480, 215, 140, 100, "Estoque OK?");

    // 4. Processo - Sucesso
    Node *nProc = Diagram_AddNode(d, SHAPE_PROCESS, 680, 235, 140, 60, "Reservar Itens");

    // 5. Subprocesso - Pagamento
    Node *nSub = Diagram_AddNode(d, SHAPE_SUBPROCESS, 880, 235, 150, 60, "Gateway Pagamento");

    // 6. Banco de Dados
    Node *nDB = Diagram_AddNode(d, SHAPE_DATABASE, 1090, 225, 130, 80, "Banco Postgres");

    // 7. Fim com Sucesso
    Node *nEnd = Diagram_AddNode(d, SHAPE_TERMINATOR, 1280, 240, 140, 50, "Pedido Entregue");

    // Rota alternativa (Não)
    Node *nWarn = Diagram_AddNode(d, SHAPE_PROCESS, 475, 380, 150, 60, "Avisar Cliente");
    if (nWarn) {
        nWarn->fillColor = (Color){ 225, 29, 72, 230 }; // Rose / Vermelho alerta
        nWarn->borderColor = (Color){ 254, 205, 211, 255 };
    }

    Node *nFail = Diagram_AddNode(d, SHAPE_TERMINATOR, 480, 490, 140, 50, "Encerrar Pedido");
    if (nFail) {
        nFail->fillColor = (Color){ 190, 18, 60, 230 };
        nFail->borderColor = (Color){ 253, 164, 175, 255 };
    }

    // Nota explicativa
    Node *nNote = Diagram_AddNode(d, SHAPE_NOTE, 880, 360, 150, 80, "Nota: Validação PIX e Cartão com webhook");

    // Conexões
    Diagram_AddConnection(d, nStart->id, nData->id, PORT_RIGHT, PORT_LEFT, "");
    Diagram_AddConnection(d, nData->id, nDec->id, PORT_RIGHT, PORT_LEFT, "");
    Diagram_AddConnection(d, nDec->id, nProc->id, PORT_RIGHT, PORT_LEFT, "Sim");
    Diagram_AddConnection(d, nDec->id, nWarn->id, PORT_BOTTOM, PORT_TOP, "Não");
    Diagram_AddConnection(d, nWarn->id, nFail->id, PORT_BOTTOM, PORT_TOP, "");
    Diagram_AddConnection(d, nProc->id, nSub->id, PORT_RIGHT, PORT_LEFT, "Aprovado");
    Diagram_AddConnection(d, nSub->id, nDB->id, PORT_RIGHT, PORT_LEFT, "Persistir");
    Diagram_AddConnection(d, nDB->id, nEnd->id, PORT_RIGHT, PORT_LEFT, "OK");

    (void)nNote;

    d->selectedNodeId = -1;
    d->selectedConnId = -1;
    Diagram_SetToast(d, "Exemplo de Fluxograma Carregado!");
}

void Templates_LoadSystemArchitecture(Diagram *d) {
    Diagram_Clear(d);

    Node *nClient = Diagram_AddNode(d, SHAPE_TERMINATOR, 100, 250, 140, 60, "Frontend Web");
    Node *nGateway = Diagram_AddNode(d, SHAPE_PROCESS, 320, 250, 150, 60, "API Gateway (C)");
    Node *nAuth = Diagram_AddNode(d, SHAPE_PROCESS, 550, 150, 150, 60, "Auth Service");
    Node *nCore = Diagram_AddNode(d, SHAPE_SUBPROCESS, 550, 340, 160, 70, "Core Engine");
    Node *nCache = Diagram_AddNode(d, SHAPE_DATABASE, 780, 150, 130, 70, "Redis Cache");
    Node *nDB = Diagram_AddNode(d, SHAPE_DATABASE, 780, 340, 130, 80, "Cluster SQL");

    Diagram_AddConnection(d, nClient->id, nGateway->id, PORT_RIGHT, PORT_LEFT, "HTTPS/JSON");
    Diagram_AddConnection(d, nGateway->id, nAuth->id, PORT_RIGHT, PORT_LEFT, "Validar JWT");
    Diagram_AddConnection(d, nAuth->id, nCache->id, PORT_RIGHT, PORT_LEFT, "Sessão");
    Diagram_AddConnection(d, nGateway->id, nCore->id, PORT_RIGHT, PORT_LEFT, "gRPC Call");
    Diagram_AddConnection(d, nCore->id, nDB->id, PORT_RIGHT, PORT_LEFT, "Queries");

    d->selectedNodeId = -1;
    d->selectedConnId = -1;
    Diagram_SetToast(d, "Arquitetura de Sistema Carregada!");
}
